### PR TITLE
apparmor: Check if apparmor_parser is available

### LINKED ISSUE #44900: Daemon does not start without apparmor
### Description

Running on Debian 11 Bullseye, and using 
`deb https://download.docker.com/linux/debian bullseye stable`
in sources.list.d.

After upgrade from version  20.10.22 to 23. results in this error

```
Error response from daemon: AppArmor enabled on system but the docker-default profile could not be loaded: running `apparmor_parser apparmor_parser --version` failed with output: 
error: exec: "apparmor_parser": executable file not found in $PATH
Error: failed to start containers: somecontainer
```

This seems to be connected with [this ticket](https://github.com/moby/moby/issues/20554) and [this](https://github.com/rancher/k3os/issues/702). There is a workaround below.

### Reproduce

1. apt update
2. apt upgrade

to install docker version 23. 

### Expected behavior

Upgrade should run smoothly.

### docker version

```bash
Client: Docker Engine - Community
 Version:           23.0.0
 API version:       1.42
 Go version:        go1.19.5
 Git commit:        e92dd87
 Built:             Wed Feb  1 17:43:17 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          23.0.0
  API version:      1.42 (minimum version 1.12)
  Go version:       go1.19.5
  Git commit:       d7573ab
  Built:            Wed Feb  1 17:43:17 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.16
  GitCommit:        31aa4358a36870b21a992d3ad2bef29e1d693bec
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client:
 Context:    default
 Debug Mode: false

Server:
 Containers: 5
  Running: 5
  Paused: 0
  Stopped: 0
 Images: 30
 Server Version: 23.0.0
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 31aa4358a36870b21a992d3ad2bef29e1d693bec
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 5.10.0-21-amd64
 Operating System: Debian GNU/Linux 11 (bullseye)
 OSType: linux
 Architecture: x86_64
 CPUs: 24
 Total Memory: 125.8GiB
 Name: server
 ID: ASHT:IIYG:NTE5:MWJH:XKZH:LCB2:RYWI:EZQM:QNT6:SNKW:AQWU:UYDW
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

### Additional Info

Workaround is to install apparmor package

`apt install apparmor`

### PR DESCRIPTION
**Pull request by vvoland**
- fixes: https://github.com/moby/moby/issues/44900
- related to: https://github.com/moby/moby/pull/42276

https://github.com/moby/moby/pull/42276 replaced the function we used to check whether AppArmor is supported on the host from the one in `runc/libcontainer/apparmor` to `containerd/pkg/apparmor`.
In containerd v1.6.1 this function has dropped a check for the `apparmor_parser` binary: https://github.com/containerd/containerd/pull/5519

It's possible in some environments that the apparmor will be enabled but the tool to load the profile is not available which will cause the `ensureDefaultAppArmorProfile` to fail completely.

This patch checks the apparmor_parser is available, if it's not, then the function returns early, but still logs a warning to the daemon log.

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: daemon/apparmor_default.go [hunk 1]
```diff
@@ -26,6 +27,11 @@ func DefaultApparmorProfile() string {
 
 func ensureDefaultAppArmorProfile() error {
 	if apparmor.HostSupports() {
+		if !aaprofile.CanInstall() {
+			logrus.Warn("AppArmor enabled on system but apparmor_parser is missing, so profile can't be loaded")
```
Comment (neersighted) [inline daemon/apparmor_default.go:31]: I don't think this error message should assert exactly what the issue is as there's two layers of indirection for the actual `LookPath` check. If we want to have informative error messages, perhaps it is better to early-return a non-fatal error if the binary is missing during subsequent attempts to actually invoke AppArmor?

**Comment by kiorky** [comment]
rel #44900

File: daemon/apparmor_default.go (hunk 1 above)
Comment (vvoland) [reply daemon/apparmor_default.go:31]: Moved the check to the `LoadProfile`.

**Comment by vvoland** [comment]
(CI failure unrelated - will go away after https://github.com/moby/moby/pull/44930 is merged)

File: daemon/apparmor_default.go [hunk 2]
```diff
@@ -8,6 +8,8 @@ import (
 
 	"github.com/containerd/containerd/pkg/apparmor"
 	aaprofile "github.com/docker/docker/profiles/apparmor"
+	"github.com/pkg/errors"
```
Comment (neersighted) [inline daemon/apparmor_default.go:11]: `pkg/errors` is deprecated -- we shouldn't use it in any new code. Let's use the stdlib instead: https://pkg.go.dev/errors#Is

File: daemon/apparmor_default.go [hunk 3]
```diff
@@ -38,7 +40,16 @@ func ensureDefaultAppArmorProfile() error {
 
 		// Load the profile.
 		if err := aaprofile.InstallDefault(defaultAppArmorProfile); err != nil {
-			return fmt.Errorf("AppArmor enabled on system but the %s profile could not be loaded: %s", defaultAppArmorProfile, err)
+			outErr := fmt.Errorf("AppArmor enabled on system but the %s profile could not be loaded: %s", defaultAppArmorProfile, err)
+
+			// Make the missing apparmor_parser a non-fatal error (to address #44900)
```
Comment (neersighted) [inline daemon/apparmor_default.go:45]: Can we make this a full URL?

File: pkg/aaparser/aaparser.go [hunk 4]
```diff
@@ -22,6 +22,13 @@ func GetVersion() (int, error) {
 	return parseVersion(output)
 }
 
+// IsAvailable checks if the apparmor_parser binary is available.
+// This should be called before any other function from this package is called.
```
Comment (neersighted) [inline pkg/aaparser/aaparser.go:26]: I'm not sure this comment line still makes sense, as `InstallDefault` is checking for the consumer.

It also feels like a `sync.Once` might make sense here, and in `GetVersion` as well.

File: profiles/apparmor/apparmor.go [hunk 5]
```diff
@@ -68,6 +71,10 @@ func macroExists(m string) bool {
 // InstallDefault generates a default profile in a temp directory determined by
 // os.TempDir(), then loads the profile into the kernel using 'apparmor_parser'.
 func InstallDefault(name string) error {
+	if !aaparser.IsAvailable() {
```
Comment (neersighted) [inline profiles/apparmor/apparmor.go:74]: I'm not sure this is where it makes sense to check -- it really seems like `generateDefault()` is overloaded/should not concern itself with asking the `apparmor_parser` for its version directly. On the other hand, an early return instead of attempting to execute the apparmor template also makes sense.

Overall, I think we're hitting a lot of code smells here, and some restructuring might be in order.

Comment (neersighted) [inline profiles/apparmor/apparmor.go:74]: I'm not sure this is where it makes sense to check -- it really seems like `generateDefault()` is overloaded/should not concern itself with asking the `apparmor_parser` for its version directly. On the other hand, an early return instead of attempting to execute the apparmor template also makes sense.

Overall, I think we're hitting a lot of code smells here, and some restructuring might be in order.

**Review by neersighted — CHANGES_REQUESTED** [review]
Some simple correctness issues, but a step in the right direction. I feel as though fundamentally this package is not well architected/abstracted and while a couple tweaks can make the initial patch good, I'd like to explore some more refactoring/restructuring to better separate concerns, calls to `apparmor_parser`, and make the error flow better.

File: profiles/apparmor/apparmor.go (hunk 5 above)
Comment (thaJeztah) [reply profiles/apparmor/apparmor.go:74]: Wondering if we should bail out before even reaching `ensureDefaultAppArmorProfile()`, instead of trying to run `aaprofile.InstallDefault`.

**Comment by thaJeztah** [comment]
>  it really seems like generateDefault() is overloaded/should not concern itself with asking the apparmor_parser for its version directly.

Even wondering if we still need the version at all. I see it was added in https://github.com/moby/moby/commit/8cf89245f5b5f9abb066f599cb69bfe0202bae5d (https://github.com/moby/moby/pull/17002) (October 2015), updated in https://github.com/moby/moby/commit/f8db9a09e0ec9b1925839ffff4f1cc5fe3ace630, and later fixed in https://github.com/moby/moby/commit/4bf7a84c969b9309b0534a61af55b8bb824acc0a (https://github.com/moby/moby/pull/20305). The original reason was described in https://github.com/moby/moby/pull/17002;

> This allows us to use the apparmor profile we have in contrib/apparmor/ and solves the problems where certain functions are not apparent on older versions of apparmor_parser on debian/ubuntu.

Which seems to be "older versions" in 2015, so very unlikely any of that still applies.

File: daemon/apparmor_default.go (hunk 2 above)
Comment (vvoland) [reply daemon/apparmor_default.go:11]: Thanks, good catch, will fix. My `gopls` always insists on the `pkg/errors` by default instead of the stdlib one and I missed this 🙄

File: pkg/aaparser/aaparser.go (hunk 4 above)
Comment (vvoland) [reply pkg/aaparser/aaparser.go:26]: I don't think `sync.Once` is appropriate here. Package containing the binary may be uninstalled after the first check was performed/or be installed after the first check happened.

**Comment by neersighted** [comment]
Notes from today's maintainer's call: We're changing this PR to be a simple reproduction of the old containerd-provided logic (just moving it upwards in the call stack) so we can be confident we're bug-for-bug compatible with the old behavior. We will then look at a refactor (probably backported to 23.0) that makes this code maintainable.

File: daemon/apparmor_default.go [hunk 6]
```diff
@@ -25,7 +33,18 @@ func DefaultApparmorProfile() string {
...
+	// Fixes: https://github.com/moby/moby/issues/44900
+	checkAppArmorParser.Do(func() {
+		if _, err := os.Stat("/sbin/apparmor_parser"); err == nil {
+			hasAppArmorParser = true
+		} else {
+			logrus.Warn("AppArmor enabled on system but \"apparmor_parser\" binary is missing, so profile can't be loaded")
+		}
+	})
+
+	if apparmor.HostSupports() && hasAppArmorParser {
```
Comment (neersighted) [inline daemon/apparmor_default.go:47]: We should move `apparmor.HostSupports()` inside the `sync.Once` as that is what the original logic did, and we additionally need to actually check apparmor is enabled in the kernel before claiming that it is in a log message.

**Review by neersighted — CHANGES_REQUESTED** [review]

File: daemon/apparmor_default.go (hunk 6 above)
Comment (thaJeztah) [reply daemon/apparmor_default.go:47]: Slightly wondering if we should put the `log.Warn` outside of the sync.Once though; maybe we want it to log on every invocation instead of only once?

Comment (neersighted) [reply daemon/apparmor_default.go:47]: I think we should just wrap twice in the `sync.Once`, that's the best way to emulate the existing behavior (only checking for the binary once at this call site) we have. Similarly, I think that it makes more sense to log inside the sync.Once, as:

1. the logging is net-new and 'good enough' to tide us over until a followup refactor
2. it allows for the most minimal change/lets us preserve the exact logic from containerd

**Comment by thaJeztah** [comment]
CI failure is a flaky test; tracked in https://github.com/moby/moby/issues/38587 (I'll give CI a kick later)

```
=== FAIL: github.com/docker/docker/pkg/plugins TestClientWithRequestTimeout (0.01s)
    client_test.go:254: assertion failed: expected an error, got nil: expected error
```

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

File: daemon/apparmor_default.go (hunk 6 above)
Comment (vvoland) [reply daemon/apparmor_default.go]: Ah right, thanks for catching this 😄 
As for logging - I think it makes sense to do it inside the `sync.Once`.
Doing it every time will simply spam the daemon log. And it could be a good thing as it definitely attracts more attention than one log line somewhere in the beginning.
However in this case it seems that in most cases where the cause of not having the `apparmor_parser` is just the characteristic of the environment rather than user's error - and he can't do anything about it.

**Review by rumpl — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
Thanks! I think @neersighted's comment has been addressed in the latest iteration, and asked @rumpl to do a review to keep me honest; I'll go ahead and merge this one 👍

**Comment by MichaIng** [comment]
Can AppArmor rules be enforced without the userland utilities installed?

I think the warning is misleading, respectively is unnecessary noise for admins which want to debug actual issues:
- The AppArmor kernel feature is usually builtin nowadays and on some distros enabled by default, like on Debian since Bullseye. One needs to explicitly disable it via `apparmor=0` command-line argument.
- But here a userspace utility is tried to be used. If the AppArmor userspace utilities are not installed, I'd take this pretty much as prove that AppArmor isn't used intentionally and a warning about this unnecessary.

A warning doesn't hurt much, but when real issues are faced and investigated, then every log entry with warning severity or above might bring admins on a wrong track. So I'd personally try to keep those at a minimum when related admin intentions are obvious.

**Comment by MichaIng** [comment]
Not sure whether I understand everything. Where shall this `apparmor=on/auto/off` be used? A Docker configuration setting? If so, to me it seems this would reintroduce the same failure on the same systems this PR is about, without admins first manually setting `apparmor=off`, while Docker would be able to be smarter and do the only correct decision automatically.

I know the background and ran into the error myself. I just think it can be perfectly silently continued without using AppArmor at all if the userland utilities, respectively `apparmor_parser`, are not installed. I cannot imagine a case where an admin of whichever distro (and whichever default kernel features/arguments) expects AppArmor to be effectively used by Docker, if the related userland utilities are not installed. So any warning about this tells the admin just what (s)he already knows, or about a feature (s)he never used and missed. So the warning doesn't help at all but is just potentially misleading noise to admins.

Another question is whether to add a section to the docs (if not there already) about the security benefits of AppArmor, how Docker is using it and how to enable it on some common distros.

**Comment by neersighted** [comment]
It's not yet determined, but probably a flag/in the daemon JSON.

The intention is to 'reintroduce the failure,' but in a deliberate way. That is to say, to force the user to specify whether or not they want AppArmor instead of silently continuing. AppArmor can be used all the way down the nested container stack, e.g. Debian containers on a Debian host. As Debian as a system is AppArmor-by-default, silently continuing without AppArmor because the user may have forgotten the package when building their Docker-in-Docker image is a rather sharp edge.

So again, this warning is just a debugging aid for the maintainers, and we are **preserving the behavior of 20.10** with this PR.

The maintainers are quite conscious of the need to make the AppArmor support better documented, solve some of the sharp edges/murky areas of the current implementation, and to attempt to make what is an import security feature **explicit in the administrator's intention** instead of attempting to guess what the user wants.

In any case, this is not the forum to discuss it -- I'd suggest that you keep an eye out for the design proposal I mentioned when it materializes. It most likely will mention/close out the linked PR, so that's probably the place to watch.

**Comment by MichaIng** [comment]
Thanks for the explanation and insights.

> So again, this warning is just a debugging aid for the maintainers

Probably then reducing severity to "debug" would be appropriate? Because

> the user may have forgotten the package when building their Docker-in-Docker image

this I think is not something you need to worry about. If anything related to AppArmor does not work but is expected to work, the first thing one would check is whether related system packages are installed, isn't it? My main concern here is that more (especially higher severity) logs can help, but can also complicate/lengthen debugging when turning focus into wrong directions. And so far I see this being the case this this warning, since it is expected (and must be ignored) in a a significant number of cases (as the reason for this PR showed).

However, 
>  this is not the forum to discuss it

so enough of my 3 (or more) cents on this topic 🙂.

**Comment by neersighted** [comment]
A warning is very appropriate as we would have used AppArmor if we had support. A warning doesn't mean something is broken necessarily; in this case we are indicating that "We would use apparmor, but you don't have the userspace binary, so we are continuing as if the kernel had it disabled" which is an appropriate usage of a warning-level log.

The absence of warnings in the logs does not imply a properly functioning install, and the presence of warnings does not imply a broken daemon. It's simply a log level meant to signify "you might want to look at this and consider if you expect it."

Moreover, log messages at the end of the day are primarily a debugging aid and are aimed more at the maintainer/administrator debugging than a user who is deploying a system. That is what documentation is for, and as you have noted, there is room for expansion.

**Comment by neersighted** [comment]
Also re: 

> And so far I see this being the case with this warning, since it is expected (and must be ignored) in a a significant number of cases (as the reason for this PR showed).

This is patently not the case; the vast majority of users **should** have AppArmor installed when their kernel supports it, and a lack of AppArmor was caused by a historical quirk of the get.docker.com script.

AppArmor in the kernel without userspace support is a unusual/more niche situation, and deserves to be noted as such to help with troubleshooting. There are very legitimate use cases for it (mixed userland with OS containers like systemd-nspawn, LXC/LXD; Docker-in-Docker) but when that happens, it is worth logging.

Moreover, the user can opt out of AppArmor (as it is only `Reccomends:` in the Docker CE packages) by using standard methods such as `apt install docker-ce apparmor-` or using APT pinning. In that case, forcing the user to also set a (hypothetical) `apparmor=off` somewhere in the Daemon config is also aligned/expected. Likewise, logging today is useful to make that situation apparent when reviewing daemon logs (as we would otherwise expect AppArmor to be present on a Debian/Ubuntu kernel reported in `docker info`).

**Comment by MichaIng** [comment]
Closing the discussion about the warning, you have some valid points there.

For the `apparmor=auto` I'd suggest this to mean "Use AppArmor if enabled and userland tools installed", so that with default config the current behaviour is preserved and AppArmor is used whenever possible but Docker does not fail if not. Since it is not uncommon, sometimes even default, to not install package recommends automatically, the `docker-ce` package install can otherwise fail (since service startup fails) without either first manually adding `apparmor=0` to cmdline or pre-creating the daemon JSON with `apparmor=off`. IMO it is a good thing to ship kernels with `apparmor=1` so that admins on any system need to install the userland tools only to get it working. The less steps to harden security, the better. Or is there a downside if the kernel feature is enabled but unused?

**Comment by neersighted** [comment]
I think that design discussion is sufficiently out of scope, so it should be addressed to the design we spitballed in our last meeting if it makes it to the proposal/implementation stage. 

However, the current thinking is that that AppArmor should be opt-out if it is available/possible, so that a missing `apparmor_parser` doesn't silently regress the defense-in-depth of the daemon. The `dind` entrypoint script/canonical methods to run nested can be updated to have more relaxed checks/rules for setting the AppArmor parameters, and users running a CentOS userland on a Ubuntu kernel (or other similar scenarios) have always had to adapt/configure things differently from the userland distributors defaults, so being secure-by-default is not an unreasonable burden.

TLDR: Building the kernel with AppArmor on by default is a statement of intent, and we should respect that statement of intent as much as possible.

**Comment by MichaIng** [comment]
Having AppArmor userland tools not installed is also a statement of intent, especially as those are a `docker-ce` recommendation (no dependency). But enough of this, now really 😄. Btw, many thanks for this PR fixing things for now 👍.

**Comment by MichaIng** [comment]
The issue has not been resolved. The error message is at a different point at container starts, but still about missing `apparmor_parser`. Looks like Docker uses the containerd function somewhere else where the same adjustment would need to be done.

In containerd it has been now reverted as well: https://github.com/containerd/containerd/pull/8086
But for v1.7. Not sure when this will be released, so whether it is worth it to further workaround this in Docker for now?

**Comment by neersighted** [comment]
Please refer to the related issues; all this information was already available, directly linked, and commenting back here just muddies the water further:

RCA and explanation: https://github.com/moby/moby/issues/44970#issuecomment-1426195907
Containerd PRs (authored by me): https://github.com/containerd/containerd/pull/8086, https://github.com/containerd/containerd/pull/8087

These PRs were specifically prepared in time for the next containerd merge window, so that we can simply re-vendor a fixed containerd instead of trying to patch around the regression, as you have suggested.

I understand you are very interested in this issue, but please be assured that the maintainers are proactive and capable, and working hard to resolve this regression.
